### ISSUE TITLE
'flags.healthbar = false' no longer works 'on joinplayer'

### ISSUE THREAD

**Issue by googolgl** [minetest/minetest#7191]
##### Issue type
 - Bug report
 
##### Minetest version
Minetest 0.5.0-dev-4827f754 (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Release
RUN_IN_PLACE=1
USE_GETTEXT=0
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=1
STATIC_SHAREDIR="."

##### OS / Hardware
Operating system:  4.13.0-37-generic #42-Ubuntu SMP Wed Mar 7 14:13:23 UTC 2018 x86_64
CPU: Intel(R) Core(TM) i7-6500U CPU @ 2.50GHz

GPU model:  VGA compatible controller: Intel Corporation HD Graphics 520 (rev 07)
OpenGL version: 

##### Summary

don't work flags.healthbar = false
Health bar don't disable

##### Steps to reproduce

 ```
 local flags = player:hud_get_flags()
  flags.healthbar = false
  player:hud_set_flags(flags)
```

**Comment by googolgl** [comment]
AND
**player:get_hp()** - don't return 1fst damage in start game.

**Comment by rubenwardy** [comment]
The code you gave is supposed to hide the health bar. What happens for you?

> player:get_hp() - don't return 1fst damage in start game.

I have no idea what this means, please may you rephrase?

**Comment by SmallJoker** [comment]
```Lua
minetest.register_on_player_hpchange(function(player, hp_change, reason)
	minetest.log("action", ("player=%s hp_change=%s new_hp=s"):format(
		player:get_player_name(), hp_change, player:get_hp() + hp_change
	))
end, false)
```
```
2018-04-04 09:54:33: ACTION[Server]: singleplayer damaged by 10 hp at (-964.54,34.215,-749.163)
2018-04-04 09:54:33: ACTION[Server]: player=singleplayer hp_change=-10 new_hp=s
```
![5 HP](https://user-images.githubusercontent.com/1497498/38295461-5dd669ca-37ee-11e8-81de-60e638b93e62.png)

Works for me. I think you forgot to apply `hp_change`, which is also given in the function:
https://github.com/minetest/minetest/blob/5070ca2111dd5b3f333a942059ef2c592de8dc24/doc/lua_api.txt#L2815

**Comment by paramat** [comment]
`print(player:get_hp())` probably returns the hp before the hp change. This can't be changed as it would break mods.

**Comment by googolgl** [comment]
Sorry, but these two different problems.
1.  I can't customize hud, because can't hide default hud of health.
And second,  **player:get_hp()**

**Comment by googolgl** [comment]
By the way, SmallJoker thanks.
Second problem is not problem.

**Comment by SmallJoker** [comment]
@googolgl The next time please open one issue for each problem. HUD flags and the player HP change callback are two completely different topics.

![grafik](https://user-images.githubusercontent.com/1497498/38353530-fe223ecc-38b6-11e8-8856-5442771c20fc.png)

```Lua
minetest.register_node(":labnode", {
	description = "bruh",
	tiles = {"default_apple.png",},
	groups = {cracky = 2},
	is_ground_content = false,
	on_punch = function(pos, node, puncher, pointed_thing)
		local flags = puncher:hud_get_flags()
		-- Swap each punch
		flags.healthbar = not flags.healthbar
		puncher:hud_set_flags(flags)
	end,
})
```
I cannot reproduce your problem. Each time I punch this node, the HP bar appears or hides.
However, I observed an issue where the HP bar data was taken to display the selected item's name (unified_inventory does that) with the result of trying to load a texture named "Apple" and "Torch". That however is closely related to #7052 (corruptions with ID=0).

**Comment by paramat** [comment]
Sorry googolgl i misunderstood your discussion of the the first issue.

**Comment by hkzorman** [comment]
I can confirm that using the following code:
```
local flags = player:hud_get_flags()
flags.healthbar = false
player:hud_set_flags(flags)
```
Doesn't works for disabling healthbar.

**Comment by Ezhh** [comment]
Tested with:

```
minetest.register_on_joinplayer(function(player)
	player:hud_set_flags({healthbar = false})
end)
```

Works in stable, but doesn't work in 5.0.0.

**Comment by Ezhh** [comment]
Throwing it in a chatcommand; however, works. So the issue here is the on_joinplayer. This change in behaviour is worrying as some mods may well break because of it.

**Comment by paramat** [comment]
Works 'on joinplayer' if run using 'minetest.after' set to 0.1s. So is something to do with trying to modify healthbar too soon in 'on joinplayer'. Singelplayer or multiplayer makes no difference.

**Comment by SmallJoker** [comment]
Partial fixes are already available: #7360 and #7348

**Comment by SmallJoker** [comment]
PRs mentioned above are merged.
